modal/_runtime/user_code_imports.py

```diff
@@ -189,6 +332,47 @@ def get_finalized_functions(
             finalized_functions[method_name] = finalized_function
         return finalized_functions
 
+    def execution_context(
+        self,
+        event_loop: UserCodeEventLoop,
+        container_io_manager: "modal._runtime.container_io_manager.ContainerIOManager",
+        function_def: api_pb2.Function,
+        is_auto_snapshot: bool,
+        call_function_callback: Optional[Callable[[dict[str, FinalizedFunction]], None]],
+    ) -> None:
+        # 1. Pre-snapshot Enter
+        if not is_auto_snapshot:
+            pre_snapshot_methods = _find_callables_for_obj(
+                self.user_cls_instance, _PartialFunctionFlags.ENTER_PRE_SNAPSHOT
+            )
+            call_lifecycle_functions(event_loop, container_io_manager, list(pre_snapshot_methods.values()))
+
+        # 2. Snapshot
+        snapshot_callback(container_io_manager, function_def)
+
+        # 3. Breakpoint wrapper
+        create_breakpoint_wrapper(container_io_manager)
+
+        # 4. Post-snapshot Enter
+        if self.user_cls_instance and not is_auto_snapshot:
+            post_snapshot_methods = _find_callables_for_obj(
+                self.user_cls_instance, _PartialFunctionFlags.ENTER_POST_SNAPSHOT
+            )
+            call_lifecycle_functions(event_loop, container_io_manager, list(post_snapshot_methods.values()))
+
+        # 5. Get Functions & Start Lifespan
+        with container_io_manager.handle_user_exception():
+            finalized_functions = self.get_finalized_functions(function_def, container_io_manager)
+
+        def exit_callback():
+            if self.user_cls_instance and not is_auto_snapshot:
+                exit_methods = _find_callables_for_obj(self.user_cls_instance, _PartialFunctionFlags.EXIT)
+                call_lifecycle_functions(event_loop, container_io_manager, list(exit_methods.values()))
+
+        _run_service_lifecycle(
```

Comment:
I was thinking this would be a Python context manager and just `yield` at this point and then just do the equivalent of `call_function_callback` in the context block of the caller (in container_entrypoint). The exit callback would also just be executed after the `yield` in this context manager then.

That would let us skip passing `call_function_callback` in here.

I think `function_def`, and `is_auto_snapshot` are directly or indirectly passed  in the Service constructors, so they don't have to be passed in again to this function and could reside as private vars on the `Service` itself